main/simulator.cpp

```diff
@@ -1,9 +1,115 @@
 #include "main/simulator.h"
 
+Simulator::Simulator(float frequency, size_t numOfRaysPerRow, int collectors,
+                     float simulationRadius, std::string_view objPath,
+                     int depth, float sourcePower)
+    : frequency_(frequency), numOfRaysPerRow_(numOfRaysPerRow),
+      manager_(SceneManager(collectors, simulationRadius)), depth_(depth),
+      sourcePower_(sourcePower) {
+  // TODO: no implementation yet
+  // model_ = manager_.getModelTriangles();
+
+  if (!manager_.loadModel(objPath)) {
+    std::stringstream ss;
+    ss << "Invalid model path";
+    throw std::invalid_argument(ss.str());
+  }
+
+  // TODO: remove
+  model_ = manager_.getReferencePlate();
+  collectors_ = manager_.getEnergyCollectors();
+
+  source_ = generators::PointSource(frequency_, numOfRaysPerRow_,
+                                    manager_.getSmapleSize(), &offsetGenerator_,
+                                    simulationRadius, sourcePower);
+}
+
 void Simulator::run() {
-  // TODO: runs simulation
+  for (size_t xIter = 0; xIter < numOfRaysPerRow_; ++xIter) {
+    for (size_t yIter = 0; yIter < numOfRaysPerRow_; ++yIter) {
+      core::Ray currentRay = source_.generateRay(xIter, yIter);
+      core::RayHitData hitData;
```

Comment:
In this particular case it is probably not an issue, but you are creating a new variable on a stack on each loop iteration. In case of heavier objects, it may cause performance drops (the best is to profile of course, but it is good to have a habit of not doing this).

Move this outside of the loop. Having this outside of the loop, is close enough to where this is first used.